### ISSUE TITLE
Animation sequence descriptions

### ISSUE THREAD

**Issue by amyjko** [wordplaydev/wordplay#470]
## What's the problem?

When using a screen reader and trying to read a `Phrase` or `Group` on screen that has a resting animation `Sequence`, there should be a description of the animation. For example, this example might be described as `dancing left and right with a small tilt`

```
Phrase('🎄' resting: (((Sequence({
  0: Pose(rotation: 0°)
  1: Pose(rotation: 15°)
  2: Pose(rotation: 0°)
  3: Pose(offset: Place(-0.25m))
  4: Pose(offset: Place(0.25m))
  5: Pose(offset: Place(-0.25m))} 0.5s)))))
```

## What's the design idea?

We could add a required description to `Sequence`, and then. use that description in `Phrase.getDescription`, `Group.getDescription`, and `Stage.getDescription` to describe the animation sequence on the output. Alternatively, we could generate a generic description based on the sequence contents, but that would probably be too low level to be helpful.

## Design specification

_Reporters can leave this blank. Designers should write a detailed, precise description of what is to be built. If images are necessary to convey this precisely, make sure to include image descriptions, so everyone can see what is in the image. Do not link to external documents; everything should be embedded here. Once the design is approved, we will remove the `needs design` tag and replace it with the `buildable` tag, signalling that it can be built._

**Comment by alissalau** [comment]
I would like to be assigned to this with @sdahl4

**Comment by sdahl4** [comment]
I would like to work on this issue @alissalau

**Comment by amyjko** [comment]
Assigned you both. This is a tricky one! Make sure to talk to people who rely on screen readers to shape this one.

**Comment by amyjko** [comment]
Great start to the design proposal! A few thoughts:

* The required input for `Sequence()` should probably be at the end, not the beginning, for backwards compatibility. Otherwise, all programs currently defining a `Sequence` would break.
* How would you automatically generate descriptions? I can't think of a way to do that. The only information is a sequence of poses, and there's no machine learning technique that exists that can reason about what those poses mean. There's also no procedural algorithm I could imagine that could derive even a templated description that would be meaningful.
* Automated description translations are already supported. Programs can be translated into other languages with an existing feature.

**Comment by sdahl4** [comment]
- That makes sense to have the input for sequence be at the end not the beginning,  here is the updated proposal for that: 
![image](https://github.com/user-attachments/assets/c3d89978-a6a1-4c0c-a790-942542e160b0)
- For introducing automated description, I would love to talk more with you about that at the meeting tomorrow, but my thoughts right now are that we would use a large language model - but with that you run into some of the same issues that I am also researching for issue #387 - where creating a large learning model based off of limited data can be more a challenge. I wonder that if creating automated descriptions is out of the scope for this issue and perhaps should be an addition for later. 
- Thank you for letting us know about translations, so that would no longer be a concern for us.

**Comment by amyjko** [comment]
Thanks @sdahl4. I think this design is close. We should talk more about the LLM idea though — in this case, there's nothing to generate a prompt to an LLM, since it has no knowledge of what the semantics of a `Sequence` or `Pose` are, and has no intelligence to infer it. It's purely a word prediction machine based on a given set of words and a large set of training data of word likelihoods. The case in which LLMs can translate code into descriptions now is when there is a large amount of code that has pre-existing descriptions in the target natural language. That's not true for Wordplay.

We can talk more tonight.

**Comment by sdahl4** [comment]
Sounds like a plan, we can talk more tonight! Is there anything I can work on perfecting with this design outside of the LLM side in the meantime? Thanks again for all the comments.

**Comment by alissalau** [comment]
### Final Proposal
@amyjko 
The design of this feature would be an extra requirement in the ‘sequence’
Instead of the original order as shown below: 

there would be an additional requirement  in the Sequence () but before the brackets {}, that indicate the map has started, called  ‘description’. This ‘description’ parameter would require a text input from the programmer that is a description of the animation set in the map.

This description would be called as Phrase.getDescription, Group.getDescription, and or Stage.getDescription to describe the animation sequence on the output. 

Our reasoning for making ‘description’ a required parameter is that it prioritizes the accessibility of all wordplay users. Additionally, it boosts the learning of all student programmers by helping standardize accessible practices into early learning. The goal of wordplay is to be an accessible programming platform, so requiring animation descriptions for visually impaired users falls in line with the ethos of this project and maximizes the experiences of all student programmers.

One possible concern is that Wordplay is mainly a tool for classrooms teaching programming. We took into consideration that required descriptions could make creating animations more complicated for younger students, but we reasoned that requiring students to complete this step could help students better formulate their ideas and reinforce learning. Students can overcome barriers by communicating with their peers and teachers if they need support in their process or just wish to share.

**Comment by amyjko** [comment]
Thank you! Design approved. Congrats @alissalau and @sdahl4.

**Comment by amyjko** [comment]
Once a quarter, I check each assignment to check the plans of the people that are assigned an issue. Please reply and indicate whether you intend to continue work on this issue. If you do not, please assign assign yourself, so we have an accurate account of who is working on an issue.

**Comment by amyjko** [comment]
@sdahl4 and @alissalau  did not reply, unassigning.

**Comment by nathanbek** [comment]
Hey @amyjko I've looked deeply into this issue and would like to get started.

**Comment by amyjko** [comment]
Great, let's do it! I think it's pretty straightforward, but let me know if you need any implementation pointers.
